src/expr.cpp

```diff
@@ -726,14 +728,66 @@ void ispc::InitSymbol(AddressInfo *ptrInfo, const Type *symType, Expr *initExpr,
                 return;
             }
 
+            // When we initialize uniform vector try to construct a vector from initializers and use a vector
+            // instruction for type conversion if needed. We can generate one instruction for type conversion when all
+            // initializers have the same atomic type. For example: const uniform float<4> f = {1.f, 2.f, 3.f, 4.f};
+            // const uniform double<4> d = {f[1], f[3], f.y, 0.0f};
+            // If it's not the case, or if initializer expression is not trivial (not of AtomicType), initialize vector
+            // per element.
+            if (symType->IsVectorType() && symType->IsUniformType() && nInits > 0) {
+                int nVectorElements = CastType<VectorType>(symType)->getVectorMemoryCount();
+                Expr *firstInit = exprList->exprs[0];
+                // Construct vector instruction only if initializer expression has AtomicType
+                if (CastType<AtomicType>(firstInit->GetType())) {
```

Comment:
Interesting observation, `CastType<>` appears to be preferable vs `isAtomic()` and `isVector()`, as it's safe for `nullptr` values.